src/piton/featurizers/featurizers.py

```diff
@@ -0,0 +1,435 @@
+
+            label_idx = 0
+            for event in patient.events:
+                while event.start > labels[label_idx].time:
+                    label_idx += 1
+                    # Create all features for label at index `label_idx`
+                    all_columns.append(
+                        [
+                            ColumnValue(self.code_to_column_index[code], count)
+                            for code, count in code_counter.items()
+                        ]
+                    )
+                    if label_idx >= len(labels):
+                        # We've reached the end of the labels for this patient,
+                        # so no point in continuing to count events past this point.
+                        # Instead, we just return the counts of all events up to this point.
+                        return all_columns
+
+                if (
+                    self.is_keep_only_none_valued_events
+                    and event.value is not None
+                ):
+                    # If we only want to keep events with no value, then skip this event
+                    # because it has a non-None value
+                    continue
+
+                for code in self.get_codes(event.code, ontology):
+                    # Increment the count for this event's code (plus any parent codes
+                    # if we are doing ontology expansion, as handled in `self.get_codes`)
+                    if code in self.included_codes:
+                        code_counter[code] += 1
+
+            if label_idx < len(labels):
+                # For all labels that occur past the last event, add all
+                # events' total counts as these labels' feature values (basically,
+                # the featurization of these labels is the count of every single event)
+                for label in labels[label_idx:]:
+                    all_columns.append(
+                        [
+                            ColumnValue(self.code_to_column_index[code], count)
+                            for code, count in code_counter.items()
+                        ]
+                    )
+        else:
+            # First, sort time bins in ascending order (i.e. [100 days, 90 days, 1 days] -> [1, 90, 100])
+            time_bins: List[datetime.timedelta] = sorted(
+                [x for x in self.time_bins if x is not None]
+            )
+
+            codes_per_bin: Dict[int, Deque[Tuple[int, datetime.datetime]]] = {
+                i: deque() for i in range(len(self.time_bins) + 1)
+            }
+
+            code_counts_per_bin: Dict[int, Dict[int, int]] = {
+                i: defaultdict(int) for i in range(len(self.time_bins) + 1)
+            }
+
+            label_idx = 0
+            for event in patient.events:
+                code: int = event.code  # type: ignore
+                while event.start > labels[label_idx].time:
+                    label_idx += 1
```

Comment:
Whenever you increment label_idx, don't you have to shuffle the code counter per bin?

Reply:
Done.